source/common/upstream/health_checker_impl.cc

```diff
@@ -23,37 +23,48 @@ class HealthCheckerFactoryContextImpl : public Server::Configuration::HealthChec
 public:
   HealthCheckerFactoryContextImpl(Upstream::Cluster& cluster, Envoy::Runtime::Loader& runtime,
                                   Envoy::Runtime::RandomGenerator& random,
-                                  Event::Dispatcher& dispatcher)
-      : cluster_(cluster), runtime_(runtime), random_(random), dispatcher_(dispatcher) {}
+                                  Event::Dispatcher& dispatcher,
+                                  const HealthCheckEventLoggerSharedPtr& event_logger)
+      : cluster_(cluster), runtime_(runtime), random_(random), dispatcher_(dispatcher),
+        event_logger_(event_logger) {}
   Upstream::Cluster& cluster() override { return cluster_; }
   Envoy::Runtime::Loader& runtime() override { return runtime_; }
   Envoy::Runtime::RandomGenerator& random() override { return random_; }
   Event::Dispatcher& dispatcher() override { return dispatcher_; }
+  const HealthCheckEventLoggerSharedPtr& eventLogger() override { return event_logger_; }
 
 private:
   Upstream::Cluster& cluster_;
   Envoy::Runtime::Loader& runtime_;
   Envoy::Runtime::RandomGenerator& random_;
   Event::Dispatcher& dispatcher_;
+  HealthCheckEventLoggerSharedPtr event_logger_;
 };
 
 HealthCheckerSharedPtr
 HealthCheckerFactory::create(const envoy::api::v2::core::HealthCheck& hc_config,
                              Upstream::Cluster& cluster, Runtime::Loader& runtime,
-                             Runtime::RandomGenerator& random, Event::Dispatcher& dispatcher) {
+                             Runtime::RandomGenerator& random, Event::Dispatcher& dispatcher,
+                             AccessLog::AccessLogManager& log_manager) {
+  HealthCheckEventLoggerSharedPtr event_logger;
+  if (!hc_config.event_log_path().empty()) {
+    event_logger =
+        std::make_shared<HealthCheckEventLoggerImpl>(log_manager, hc_config.event_log_path());
```

Comment:
Two questions about the event logger's shared_ptr:

1. ~It seems that each health checker is the sole owner of the event logger, as each health checker construction is preceded by a event logger construction. Is the reason to make is a shared ptr because below we pass the event logger to a factory context and that context can share the event logger to more owners?~ Ah I see, the ownership is shared between the health checker and all the active sessions. Why wouldn't the healthchecker own the event logger and give a reference to the active sessions? Or have the active session write to the `parent_`'s `event_logger_`?

2. Why do we return the event logger's shared ptr from the context, and pass it into constructors by reference?

Reply:
Gosh, I can't remember, I kind of just fixed a couple signatures and then propagated necessary changes until everything compiled. But you're right that it all feels weird. I'll refactor so it makes more sense. I think I only had shared_ptrs in the first place because that's how the outlier loggers were structured..